src/pydicom/dataset.py

```diff
@@ -1032,6 +1004,12 @@ def __getitem__(self, key: "slice | TagType") -> "Dataset | DataElement":
             # Not converted from raw form read from file yet; do so now
             self[tag] = DataElement_from_raw(elem, character_set, self)
 
+            # On initial read of the dataset, propagate the pixel representation
+            #   (if any) to child datasets in any sequences.
+            # This is used as part of the ambiguous VR correction for US or SS
```

Comment:
I may be misunderstanding something, but what happens if some data elements (e.g. Sequences and their child Datasets) are used in multiple Datasets?  For example, someone could create a compressed and a non-compressed version of the same Dataset.  Couldn't there be conflict over the child settings?

Reply:
I think potentially there could be conflict, it depends how deep the ambiguous element is. 

I've actually been in two minds about this (I actually like the relational attributes, I dislike the maintenance overhead). I've been looking to see if there's an alternative to the current system and haven't had any luck, because I think fundamentally you need a reference to the parent somewhere in the child.

I can close this and give the relational attributes a hard stare instead to see if there's anything we're overlooking? One thing I think we potentially might want is to add code to guard against circular references:

```python
ds = Dataset()
ds.BeamSequence = [ds]  # !
```

Another is that `Dataset.__delitem__` doesn't remove references to the parent from the children (the children can still exist but are no longer part of the parent, but still point to that original parent).

Reply:
I'm still (naively, probably) wondering about whether a top-down approach could work - e.g. only handle ambiguous VR from the 'root' `Dataset`.  Certainly a pass could be done on the whole thing before writing, and after reading from a file, the trick is how to do it if the user is changing elements in a Dataset.  I'm thinking in terms of just setting the element's VR rather than storing any kind of tree or parent/child relationship.  Not sure if that is possible or if it just runs into the same problems anyway.

Also, I think my comment applies to current code anyway, if someone added a Sequence reference to two different `Dataset`s, the parent would get changed on the second assignment.  So maybe we could still go forward with an improvement on that.

I'm still trying to think it all through ...

Reply:
The main issue is, I think, that on read with implicit VR the ambiguous element is a RawDataElement in what may be a child dataset in a sequence. And because the ambiguous VR correction in this case is done in `Dataset.__getitem__()` at the element level (unlike the VR correction on writing which is done from the root dataset down), there must be a need to work back up to the dataset root.

Reply:
> One solution is on read for implicit VR to selectively convert RawDataElement to DataElement for Pixel Representation and the ambiguous VR elements, iterating down into sequences to find any. Which can be done, but at a slightly longer read time.

That sounds promising, although its late (for me) now so I'll need tomorrow's brainpower to think about that...

This triggers another thought though - can you `asv run` bench_nested_seq.py on this PR code vs main? (or I could give it a go tomorrow)  I don't think it should happen here from what I see in the code, but we don't want to regress too much on the speed improvements from #1734.  And OTOH, 'iterating down sequences' mentioned above could actually risk that.

Reply:
`dev-tree`
```
[12.50%] ··· Running (bench_nested_seq.TimeNestedSeqAccess.time_index_to_nested_items--)...
[62.50%] ··· ...h_nested_seq.TimeNestedSeqAccess.time_index_to_nested_items          222±0.6ms
[75.00%] ··· bench_nested_seq.TimeNestedSeqAccess.time_iterate_nested_elems            171±1ms
[87.50%] ··· bench_nested_seq.TimeNestedSeqAccess.time_iterate_preload_tags           87.4±1ms
[100.00%] ··· bench_nested_seq.TimeNestedSeqAccess.track_len_top_sequence                 20000
```

`main`
```
[12.50%] ··· Running (bench_nested_seq.TimeNestedSeqAccess.time_index_to_nested_items--)...
[62.50%] ··· ...h_nested_seq.TimeNestedSeqAccess.time_index_to_nested_items            222±1ms
[75.00%] ··· bench_nested_seq.TimeNestedSeqAccess.time_iterate_nested_elems            170±1ms
[87.50%] ··· bench_nested_seq.TimeNestedSeqAccess.time_iterate_preload_tags           85.8±1ms
[100.00%] ··· bench_nested_seq.TimeNestedSeqAccess.track_len_top_sequence                 20000
```

These were typical of the results of 3 runs of the benchmarking after ignoring the first run

Reply:
The results can be quite noisy between runs, but they appear pretty similar:
```
[ 0.00%] · For pydicom commit fd4e9c7a <main>:
[ 0.00%] ·· Benchmarking virtualenv-py3.10-numpy
[ 6.25%] ··· Running (bench_nested_seq.TimeNestedSeqAccess.time_index_to_nested_items--)...
[31.25%] ··· ...h_nested_seq.TimeNestedSeqAccess.time_index_to_nested_items            221±1ms
[37.50%] ··· bench_nested_seq.TimeNestedSeqAccess.time_iterate_nested_elems          168±0.7ms
[43.75%] ··· bench_nested_seq.TimeNestedSeqAccess.time_iterate_preload_tags           86.9±2ms
[50.00%] ··· bench_nested_seq.TimeNestedSeqAccess.track_len_top_sequence                 20000
```
```
[50.00%] · For pydicom commit 76cef63e <dev-tree>:
[50.00%] ·· Building for virtualenv-py3.10-numpy..
[50.00%] ·· Benchmarking virtualenv-py3.10-numpy
[56.25%] ··· Running (bench_nested_seq.TimeNestedSeqAccess.time_index_to_nested_items--)...
[81.25%] ··· ...h_nested_seq.TimeNestedSeqAccess.time_index_to_nested_items          220±0.9ms
[87.50%] ··· bench_nested_seq.TimeNestedSeqAccess.time_iterate_nested_elems            169±2ms
[93.75%] ··· bench_nested_seq.TimeNestedSeqAccess.time_iterate_preload_tags           90.6±1ms
[100.00%] ··· bench_nested_seq.TimeNestedSeqAccess.track_len_top_sequence                 20000
```

Reply:
Thanks, that looks good.  I'll try to review the code properly in the next couple of days.

> The benchmark only uses `DataElement` elements, too. After read they'd normally all be `RawDataElement`.

I think for the speed testing purposes here the type shouldn't matter much, but if there's a reason to think about it we could add something to the benchmark.